## Importamos librerias junto con el Dataset

In [2]:
import numpy as np
from sklearn.datasets import load_wine
from sklearn.model_selection import KFold, StratifiedKFold, cross_val_score, train_test_split, LeaveOneOut
from sklearn.linear_model import LogisticRegression, LinearRegression
from sklearn.metrics import confusion_matrix, accuracy_score, precision_score, recall_score, f1_score, roc_auc_score, mean_squared_error, mean_absolute_error
from sklearn.preprocessing import Binarizer
from sklearn.multioutput import MultiOutputRegressor

## Cargamos el Dataset de vinos

In [4]:
wine = load_wine()
X, y = wine.data, wine.target

In [5]:
# Convertir las etiquetas a formato binario para ciertas métricas
y_binary = (y == y[0]).astype(int)

# Definir los modelos
log_reg = LogisticRegression(max_iter=10000)
linear_reg = LinearRegression()

## Aplicamos las metricas de evaluacion y tecnicas de validacion cruzada

In [7]:
# Técnicas de Validación Cruzada
kf = KFold(n_splits=5, shuffle=True, random_state=42)
loo = LeaveOneOut()
skf = StratifiedKFold(n_splits=5, shuffle=True, random_state=42)

# K-Fold Cross-Validation
kf_scores = cross_val_score(log_reg, X, y, cv=kf, scoring='accuracy')

# Leave-One-Out Cross-Validation
loo_scores = cross_val_score(log_reg, X, y, cv=loo, scoring='accuracy')

# Stratified K-Fold Cross-Validation
skf_scores = cross_val_score(log_reg, X, y, cv=skf, scoring='accuracy')

# Métricas de Evaluación
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
log_reg.fit(X_train, y_train)
y_pred = log_reg.predict(X_test)

conf_matrix = confusion_matrix(y_test, y_pred)
accuracy = accuracy_score(y_test, y_pred)
precision = precision_score(y_test, y_pred, average='weighted')
recall = recall_score(y_test, y_pred, average='weighted')
f1 = f1_score(y_test, y_pred, average='weighted')
# Para el AUC-ROC, necesitamos predecir las probabilidades
if len(np.unique(y)) == 2:  # AUC-ROC sólo tiene sentido para problemas binarios
    auc_roc = roc_auc_score(y_test, log_reg.predict_proba(X_test)[:, 1])
else:
    auc_roc = 'N/A'

# Para las métricas de regresión usamos el modelo de regresión lineal
linear_reg.fit(X_train, y_train)
y_pred_reg = linear_reg.predict(X_test)
mse = mean_squared_error(y_test, y_pred_reg)
mae = mean_absolute_error(y_test, y_pred_reg)

# Imprimir resultados
print(f"K-Fold Cross-Validation Accuracy: {kf_scores.mean()}")
print(f"Leave-One-Out Cross-Validation Accuracy: {loo_scores.mean()}")
print(f"Stratified K-Fold Cross-Validation Accuracy: {skf_scores.mean()}")
print("Matriz de Confusión:")
print(conf_matrix)
print(f"Exactitud (Accuracy): {accuracy}")
print(f"Precisión (Precision): {precision}")
print(f"Exhaustividad (Recall): {recall}")
print(f"Puntuación F1 (F1 Score): {f1}")
print(f"AUC-ROC: {auc_roc}")
print(f"Error Cuadrático Medio (MSE): {mse}")
print(f"Error Absoluto Medio (MAE): {mae}")

K-Fold Cross-Validation Accuracy: 0.9552380952380952
Leave-One-Out Cross-Validation Accuracy: 0.9550561797752809
Stratified K-Fold Cross-Validation Accuracy: 0.9663492063492063
Matriz de Confusión:
[[14  0  0]
 [ 0 14  0]
 [ 0  0  8]]
Exactitud (Accuracy): 1.0
Precisión (Precision): 1.0
Exhaustividad (Recall): 1.0
Puntuación F1 (F1 Score): 1.0
AUC-ROC: N/A
Error Cuadrático Medio (MSE): 0.06853348464256068
Error Absoluto Medio (MAE): 0.20304725956613046


# Resultados

**K-Fold Cross-Validation Accuracy**
- **Valor:** 0.9552380952380952
- **Significado:** La precisión promedio obtenida mediante validación cruzada K-Fold. La precisión mide la proporción de predicciones correctas. Un valor de aproximadamente 0.955 indica que el modelo clasifica correctamente el 95.5% de las veces en el conjunto de datos de entrenamiento.

**Leave-One-Out Cross-Validation Accuracy**
- **Valor:** 0.9550561797752809
- **Significado:** La precisión promedio utilizando validación Leave-One-Out (LOO). En LOO, se usa cada punto de datos como un conjunto de prueba mientras se entrena el modelo en el resto de los datos. Un valor de aproximadamente 0.955 indica que el modelo clasifica correctamente el 95.5% de las veces.

**Stratified K-Fold Cross-Validation Accuracy**
- **Valor:** 0.9663492063492063
- **Significado:** La precisión promedio obtenida mediante validación cruzada K-Fold estratificada. La precisión mide la proporción de predicciones correctas. Un valor de aproximadamente 0.966 indica que el modelo clasifica correctamente el 96.6% de las veces, lo que muestra un buen desempeño del modelo.

**Matriz de Confusión**
- **Valor:** 
            ```
             [[14 0 0],
             [ 0 14 0],
             [ 0 0 8]],
             ```
- **Significado:**
- 14: Verdaderos negativos para la primera clase (casos negativos correctamente predichos).
- 0: Falsos positivos para la primera clase (casos negativos incorrectamente predichos como positivos).
- 0: Falsos negativos para la primera clase (casos positivos incorrectamente predichos como negativos).
- 14: Verdaderos positivos para la segunda clase (casos positivos correctamente predichos).
- 0: Falsos positivos para la segunda clase (casos negativos incorrectamente predichos como positivos).
- 0: Falsos negativos para la segunda clase (casos positivos incorrectamente predichos como negativos).
- 8: Verdaderos positivos para la tercera clase (casos positivos correctamente predichos).

**Exactitud (Accuracy)**
- **Valor:** 1.0
- **Significado:** La proporción de predicciones correctas sobre el total de predicciones. Un valor de 1.0 indica que el modelo es correcto el 100% del tiempo, mostrando un desempeño perfecto en el conjunto de prueba.

**Precisión (Precision)**
- **Valor:** 1.0
- **Significado:** La precisión es la proporción de verdaderos positivos sobre el total de positivos predichos (verdaderos positivos + falsos positivos). Un valor de 1.0 indica que cuando el modelo predice positivo, es correcto el 100% del tiempo.

**Exhaustividad (Recall)**
- **Valor:** 1.0
- **Significado:** La exhaustividad o recall es la proporción de verdaderos positivos sobre el total de positivos reales (verdaderos positivos + falsos negativos). Un valor de 1.0 indica que el modelo captura el 100% de los casos positivos.

**Puntuación F1 (F1 Score)**
- **Valor:** 1.0
- **Significado:** La puntuación F1 es la media armónica de la precisión y el recall. Un valor de 1.0 indica un equilibrio perfecto entre precisión y recall.

**AUC-ROC**
- **Valor:** N/A
- **Significado:** El área bajo la curva ROC mide la capacidad del modelo para distinguir entre clases. No se calculó ya que el problema no es binario en este caso.

**Error Cuadrático Medio (MSE)**
- **Valor:** 0.06853348464256068
- **Significado:** El MSE mide la media de los errores al cuadrado entre los valores reales y los valores predichos. Un valor menor indica un mejor desempeño en términos de error cuadrático. En este caso, un MSE de aproximadamente 0.068 indica un error cuadrático medio bajo.

**Error Absoluto Medio (MAE)**
- **Valor:** 0.20304725956613046
- **Significado:** El MAE mide la media de los errores absolutos entre los valores reales y los valores predichos. Un valor menor indica un mejor desempeño en términos de error absoluto. En este caso, un MAE de aproximadamente 0.203 indica un error absoluto medio bajo.
